crates/sui-core/src/consensus_throughput_calculator.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::authority::AuthorityMetrics;
+use arc_swap::ArcSwap;
+use narwhal_types::TimestampMs;
+use parking_lot::Mutex;
+use std::collections::{BTreeMap, VecDeque};
+use std::num::NonZeroU64;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering::Relaxed;
+use std::sync::Arc;
+use tracing::{info, warn};
+
+const DEFAULT_OBSERVATIONS_WINDOW: u64 = 120; // number of observations to use to calculate the past traffic
+const DEFAULT_TRAFFIC_PROFILE_UPDATE_WINDOW_SECS: u64 = 60; // seconds that need to pass between two consecutive traffic profile updates
+const DEFAULT_TRAFFIC_PROFILE_COOL_DOWN_THROUGHPUT_THRESHOLD: u64 = 10; // 10% of throughput
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct TrafficProfile {
+    pub level: Level,
+    /// The lower range of the throughput that this profile is referring to. For example, if
+    /// `throughput = 1_000`, then for values >= 1_000 this traffic profile applies.
+    pub throughput: u64,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub enum Level {
+    Low,
+    High,
+}
+
+impl Level {
+    fn as_int(&self) -> usize {
+        match self {
+            Level::Low => 0,
+            Level::High => 1,
+        }
+    }
+}
+
+pub struct TrafficProfileRanges {
+    /// Holds the traffic profiles by the traffic range (upper_throughput, cool_down_threshold)
+    profiles: BTreeMap<u64, TrafficProfile>,
+}
+
+impl TrafficProfileRanges {
+    pub fn new(profiles: &[TrafficProfile]) -> Self {
+        let mut p: BTreeMap<u64, TrafficProfile> = BTreeMap::new();
+
+        for profile in profiles {
+            assert!(
+                !p.iter().any(|(_, pr)| pr.level == profile.level),
+                "Attempted to insert profile with same level"
+            );
+            assert!(
+                p.insert(profile.throughput, *profile).is_none(),
+                "Attempted to insert profile with same throughput"
+            );
+        }
+
+        Self { profiles: p }
+    }
+
+    pub fn lowest_profile(&self) -> TrafficProfile {
+        *self
+            .profiles
+            .first_key_value()
+            .expect("Should contain at least one traffic profile")
+            .1
+    }
+    /// Resolves the traffic profile that corresponds to the provided throughput.
+    pub fn resolve(&self, current_throughput: u64) -> TrafficProfile {
+        let mut iter = self.profiles.iter();
+        while let Some((threshold, profile)) = iter.next_back() {
+            if current_throughput >= *threshold {
+                return *profile;
+            }
+        }
+        panic!("{}", format!("Method should always be able to detect a traffic profile to cover the provided throughput {}, {:?}", current_throughput, self.profiles));
```

Comment:
is there no way to architect this code in a way that we default to some conservative parameters (eg LOW) rather than panic in case something goes wrong?

Reply:
Yeah I am not very proud of that panic here - and initially I had it falling back to a default, but that should not really happen. Basically all we need to make sure that we'll never panic is making sure that we have at list one profile with throughput `0`. Now, returning one profile in any case we could do.